File: backend/app/schemas/match.py
from pydantic import BaseModel


class MatchRequest(BaseModel):
    focus_area: str
    location: str
    budget: str
    timeline: str
    expected_outcomes: str


class ComparisonField(BaseModel):
    label: str
    matched: bool


class NGOMatch(BaseModel):
    name: str
    match_score: float
    expertise: list[str]
    regions: list[str]
    beneficiaries: list[str]
    capabilities: list[str]
    past_projects: list[str]
    email: str
    phone: str
    address: str
    comparison: list[ComparisonField]


class MatchResponse(BaseModel):
    matches: list[NGOMatch]


class SuggestionsRequest(BaseModel):
    requirement: MatchRequest
    ngo_name: str
    ngo_expertise: list[str]
    ngo_regions: list[str]
    ngo_beneficiaries: list[str]
    ngo_capabilities: list[str]
    ngo_past_projects: list[str]


class SuggestionsResponse(BaseModel):
    suggestions: list[str]

File: src/api/match.ts
import { postJson } from "./client";
import type { ComparisonField, MatchRequest, NGOMatch } from "../types/match";

interface ComparisonFieldBody {
  label: string;
  matched: boolean;
}

interface NGOMatchResponseBody {
  name: string;
  match_score: number;
  expertise: string[];
  regions: string[];
  beneficiaries: string[];
  capabilities: string[];
  past_projects: string[];
  email: string;
  phone: string;
  address: string;
  comparison: ComparisonFieldBody[];
}

interface MatchResponseBody {
  matches: NGOMatchResponseBody[];
}

function toComparisonField(body: ComparisonFieldBody): ComparisonField {
  return { label: body.label, matched: body.matched };
}

function toNGOMatch(body: NGOMatchResponseBody): NGOMatch {
  return {
    name: body.name,
    matchScore: body.match_score,
    expertise: body.expertise,
    regions: body.regions,
    beneficiaries: body.beneficiaries,
    capabilities: body.capabilities,
    pastProjects: body.past_projects,
    email: body.email,
    phone: body.phone,
    address: body.address,
    comparison: body.comparison.map(toComparisonField),
  };
}

export async function findMatches(request: MatchRequest): Promise<NGOMatch[]> {
  const body = await postJson<MatchResponseBody>("/api/match", {
    focus_area: request.focusArea,
    location: request.location,
    budget: request.budget,
    timeline: request.timeline,
    expected_outcomes: request.expectedOutcomes,
  });
  return body.matches.map(toNGOMatch);
}

interface SuggestionsResponseBody {
  suggestions: string[];
}

export async function getNgoSuggestions(
  match: NGOMatch,
  requirement: MatchRequest,
): Promise<string[]> {
  const body = await postJson<SuggestionsResponseBody>("/api/match/suggestions", {
    requirement: {
      focus_area: requirement.focusArea,
      location: requirement.location,
      budget: requirement.budget,
      timeline: requirement.timeline,
      expected_outcomes: requirement.expectedOutcomes,
    },
    ngo_name: match.name,
    ngo_expertise: match.expertise,
    ngo_regions: match.regions,
    ngo_beneficiaries: match.beneficiaries,
    ngo_capabilities: match.capabilities,
    ngo_past_projects: match.pastProjects,
  });
  return body.suggestions;
}
